### ISSUE TITLE
Adapters should be Rack apps not middleware

### ISSUE THREAD

**Issue by mislav** [lostisland/faraday#47]
This is our current Rack app (i.e. the "inner app" wrapped by middleware):

```
app = lambda { |env|
  response = Response.new
  response.finish(env) unless env[:parallel_manager]
  env[:response] = response
}
```

It's really kind of a hack. I think our adapters should each perform this and stop acting like middleware when they really play a role of an endpoint. I can't think of a use case where you would want to mount middleware _after_ the adapter.

**Comment by dentarg** [comment]
@iMacTia also not closed by #750, fyi

**Comment by iMacTia** [comment]
@dentarg I agree, #750 only provides a "workaround" to this, but Adapters are now forced to be at the end of the stack as they were endpoints. The result is basically the same as the one suggested in this ticket, even though they kept their "middleware" nature. I considered the refactoring but it was causing other issues, plus it would require all adapters (even external ones out of our control like the Typhoeus one) to be changed
